rfcs/20221124-monorepo.md

```diff
@@ -0,0 +1,178 @@
+
+The [integration](#step-4-integration) section in the latter part of the document describes more interesting outcomes made possible by the monorepo. They may or may not be exclusive to monorepos but they're surely made easier by it.
+
+## Migration path
+
+First we will merge the API and the frontend. This decision was made for the following reasons.
+
+1. API and frontend are tightly linked. The frontend is a direct consumer of what the API produces.
+
+1. The API and frontend form the "service" side of Openverse that directly faces the users (both API consumers and Search engine users).
+
+1. The frontend uses ECS deployments and the API is well on the same track. This makes it possible for them to share some deployment code.
+
+1. To the RFC author, the API and frontend are very familiar so merging them would be easier. Adding a third component would make the task daunting.
+
+1. Merging incurs a productivity hit for the initial transition. So merging everything in one swoop is not ideal.
+
+1. The API’s comprehensive tooling for developer documentation can benefit frontend devs and create a unified docs site for contributors.
+
+1. The API is already organised by stack folders so the `frontend/` directory will fit right in with the others like `api/` and `ingestion_server/`.
+
+1. The API and frontend share identical tooling for Git hooks, linting and formatting. We will fight our tools less and encounter minimal friction.
+
+   - In fact, we employ a number of hacks to install and configure pre-commit for the frontend. Merging it with the API eliminates the need for such hacks.
+
+1. The entire system can be integration tested during releases. The real API, populated with test data, can even replace the Talkback server.
+
+The `WordPress/openverse-api` repo will absorb the `WordPress/openverse-frontend` repo. The `WordPress/openverse-catalog` will also be merged, _later_.
+
+### Reference
+
+I'm following the steps listed below in a fork at [@dhruvkb/monopenverse](https://github.com/dhruvkb/monopenverse/). You can refer to the fork, but note that it is a comes from a place of haste and has not been treated with the same level of love and care that the final treatment will receive.
+
+### Step 0: Prerequisites
+
+#### Get the timing right
+
+The first step will be to release the frontend, call a code freeze and pause work on it. This is to prevent the frontend repo from continuing to drift as we merge a snapshot of it with the API.
+
+This can prove difficult given how productive our team is, so we will need to channel this productivity towards the catalog in the meantime. I can foresee the end-to-end migration taking one week (ideal scenario) to one fortnight (worst case scenario).
+
+### Step 1: Merge with histories
+
+This is a quick process.
+
+1. Move the entire content of frontend inside a `frontend/` directory, except the following top-level files and folders. Please comment if you can add to this list.
```

Comment:
What is the plan for making `pnpm` work in the monorepo? Do we want to go ahead and making `pnpm` live at the root of the repository and treat the frontend as a `pnpm` monorepo package? https://pnpm.io/workspaces

I think this would be necessary for us to be able to (reasonably) merge the browser extension and the JS that is currently in WordPress/openverse into the monorepo... otherwise we'll be fighting weird fights with pnpm in different directories :thinking:

Reply:
That was an issue. Currently I had everything working with `pnpm` with `package.json` moved into the `frontend/` directly (no package manager at the root), but I definitely should look into the `workspaces` feature.